File: Java/DS/Doubly linked list/Delete.java
import java.util.Scanner;

class Node{
    Node prev;
    int data;
    Node next;
}

class Delete{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        Node head = null, tail = null;
        System.out.println("1. add 2. display 3. exit");
        do{
            System.out.print("Enter choice : ");
            int choice = sc.nextInt();
            if(choice == 1){
                System.out.print("Enter value : ");
                int value = sc.nextInt();
                if(head == null){
                    tail = addLast(tail, value);
                    head = tail;
                }
                else
                    tail = addLast(tail, value);
            }
            else if(choice == 2){
                printList(head);
                revPrintList(tail);
            }
            else if(choice == 3){
                break;
            }
        }while(true);

        do{
            System.out.print("Enter key : ");
            int key = sc.nextInt();
            if(key == -1)
                break;
            else if(head == null && tail == null){
                System.out.println("Empty");
                break;
            }
            else if(head.next == null){
                head = tail = null;
                break;
            }
            else if(head.data == key){
                head = head.next;
                head.prev = null;
            }
            else if(tail.data == key){
                tail = tail.prev;
                tail.next = null;
            }
            else{
                head = delete(head, key);
            }
            printList(head);
            revPrintList(tail);
        }while(true);


        sc.close();
    }

    public static Node delete(Node head, int key){
        Node temp = head;
        while(temp.next != null){
            if(temp.next.data == key){
                temp.next.next.prev = temp;
                temp.next = temp.next.next;
                return head;
            }
            temp = temp.next;
        }
        System.out.println("Key not found.");
        return head;
    }

    public static void search(Node head, int key){
        Node temp = head;
        while(temp != null){
            if(temp.data == key){
                System.out.print("Found\n");
                return;
            }
            temp = temp.next;
        }
        System.out.print("Not found\n");
    }

    public static Node addFirst(Node head, int value){
        Node temp = new Node();
        temp.data = value;
        if(head == null){
            temp.prev = null;
            temp.next = null;
        }
        else{
            temp.prev = null;
            temp.next = head;
            head.prev = temp;
        }
        return temp;
    }

    public static Node addLast(Node tail, int value){
        Node temp = new Node();
        temp.data = value;
        temp.next = null;
        if(tail == null){
            temp.prev = null;
        }
        else{
            temp.prev = tail;
            tail.next = temp;
        }
        return temp;
    }

    public static void printList(Node head){
        Node temp = head;
        while(temp != null){
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void revPrintList(Node tail){
        Node temp = tail;
        while(temp != null){
            System.out.print(temp.data + " ");
            temp = temp.prev;
        }
        System.out.println();
    }
}


// 1 1 1 2 1 3 1 4 1 5 1 6 1 7 1 8 1 9 1 10

File: C language/DS/Doubly linkedList/Delete.c
#include<stdio.h>
#include<stdlib.h>

struct node{
    struct node *prev;
    int data;
    struct node *next;
};

void printList(struct node *head){
    struct node *temp = head;
    while(temp != NULL){
        printf("%d ", temp -> data);
        temp = temp -> next;
    }
    printf("\n");
}

void revPrintList(struct node *tail){
    struct node *temp = tail;
    while(temp != NULL){
        printf("%d ", temp -> data);
        temp = temp -> prev;
    }
    printf("\n");
}

void Deallocate(struct node *head){
    struct node *temp = head;
    struct node *hold;
    while(temp != NULL){
        hold = temp -> next;
        free(temp);
        temp = hold;
    }
}

struct node * getFirst(struct node *head, int value){
    struct node *temp = malloc(sizeof(struct node));
    temp -> data = value;
    temp -> prev = NULL;

    if(head == NULL){
        temp -> next = NULL;
    }
    else{
        temp -> next = head;
        head -> prev = temp;
    }

    return temp;
}

struct node * addLast(struct node *tail, int value){
    struct node *temp = malloc(sizeof(struct node));
    temp -> data = value;
    temp -> next = NULL;

    if(tail == NULL){
        temp -> prev = NULL;
    }
    else{
        tail -> next = temp;
        temp -> prev = tail;
    }

    return temp;
}

void search(struct node *head, int key){
    struct node *temp = head;
    while(temp != NULL){
        if(temp -> data == key){
            printf("Found\n");
            return;
        }
        temp = temp -> next;
    }
    printf("Not found\n");
}

struct node * delete(struct node *head, int key){
    struct node *temp = head;
        while(temp -> next != NULL){
            if(temp -> next -> data == key){
                temp -> next -> next -> prev = temp;
                temp -> next = temp -> next -> next;
                return head;
            }
            temp = temp -> next;
        }
        printf("Key not found.\n");
        return head;
}

int main(){
    int choice, value, key;
    struct node *head = NULL, *tail = NULL;

    printf("1. add 2. display 3. exit\n");

    do{
        printf("Enter choice : ");
        scanf("%d",&choice);
        if(choice == 1){
            printf("Enter the value : ");
            scanf("%d",&value);
            if(head == NULL){
                tail = addLast(tail, value);
                head = tail;
            }
            else{
                tail = addLast(tail, value);
            }
        }
        else if(choice == 2){
            printList(head);
            revPrintList(tail);
        }
        else if(choice == 3){
            break;
        }
    }while(1);

    do{
        printf("Enter the key : ");
        scanf("%d",&key);
        if(key == -1){
            break;
        }

        else if(head == NULL && tail == NULL){
            printf("Empty.\n");
            break;
        }

        else if(head -> next == NULL){
            head = tail = NULL;
        }

        else if(head -> data == key){
            head = head -> next;
            head -> prev = NULL;
        }

        else if(tail -> data == key){
            tail = tail -> prev;
            tail -> next = NULL;
        }

        else{
            head = delete(head, key);
        }

        printList(head);
        revPrintList(tail);

    }while(1);

    Deallocate(head);
}
